File: problems/Binary Tree Right Side View/rightSideView.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<int> RightSideView(TreeNode root) {
        var result = new List<int>();
        var bfs = new Queue<TreeNode>();

        if (null != root) {
            bfs.Enqueue(root);
        }

        while (0 < bfs.Count) {
            var count = bfs.Count;
            TreeNode node = null;

            for (var i = 0; count > i; ++i) {
                node = bfs.Dequeue();

                if (null != node.left) {
                    bfs.Enqueue(node.left);
                }
                if (null != node.right) {
                    bfs.Enqueue(node.right);
                }
            }

            result.Add(node.val);
        }

        return result;
    }
}

File: problems/Binary Tree Right Side View/rightSideView.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<Integer>();
        Queue<TreeNode> bfs = new LinkedList<TreeNode>();

        if (null != root) {
            bfs.add(root);
        }

        while (0 < bfs.size()) {
            int count = bfs.size();
            TreeNode node = null;

            while (0 < count--) {
                node = bfs.poll();

                if (null != node.left) {
                    bfs.add(node.left);
                }

                if (null != node.right) {
                    bfs.add(node.right);
                }
            }

            result.add(node.val);
        }

        return result;
    }
}
